### ISSUE TITLE
Writing Your Own Toy Compiler Using Flex Bison and LLVM (gnuu.org)

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5736]
Writing Your Own Toy Compiler Using Flex, Bison and LLVM (gnuu.org)<br>
<br>
https://ift.tt/2FVHMtp<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<p><strong>Update (March 19 2010)</strong>: this article was updated for <strong>LLVM 2.6</strong> thanks to a great patch by <a href="https://github.com/ashgti/my_toy_compiler">John Harrison</a>. He rocks!</p>
<p>I&rsquo;ve always been interested in compilers and languages, but interest only gets you so far. A lot of the concepts of compiler design can easily go way over most programmers&rsquo; heads, even the intelligent ones. Needless to say, I&rsquo;ve tried, without much success, to write a small toy language/compiler before. I&rsquo;d usually get caught up at the semantic parsing stage. And again, needless to say, this post is mostly inspired by my latest attempt, though this one has been much more successful (so far).</p>
<p>Fortunately over the last few years I&rsquo;ve been involved in <a href="http://yard.soen.ca">some</a> <a href="http://www.eecs.ucf.edu/~leavens/JML//index.shtml">projects</a> that helped give me perspective and experience on what&rsquo;s really involved in building a compiler. The other thing I&rsquo;ve been lucky to have in my corner this time is the help of <a href="https://llvm.org">LLVM</a>, a tool which I&rsquo;m hardly qualified to talk too much about, but it&rsquo;s been quite handy in implementing most of the business end (read: complex aspects) of my toy compiler.</p>
<h2>So Why Are You Reading This?</h2>
<p>Maybe you want to see what I&rsquo;ve been doing with my time. It&rsquo;s more likely, however, that you&rsquo;re probably interested in compilers and languages as I am, and have probably been hitting similar roadblocks. You&rsquo;ve probably wanted to try this but never found the resources, or did but couldn&rsquo;t quite follow. The goal of this article is to provide such a resource and explain in a relatively step by step manner how to create the most basic-but-functional compiler from start to &ldquo;finish&rdquo;.</p>
<p>I won&rsquo;t be covering much theory, so if you haven&rsquo;t brushed up on your <a href="https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_Form">BNF grammars</a>, <a href="https://en.wikipedia.org/wiki/Abstract_syntax_tree">AST data structures</a> and the basic <a href="https://en.wikipedia.org/wiki/Compiler">compiler pipeline</a>, I suggest you do so. That said, I plan on keeping this as simple as possible. The goal, of course, is to make this an easy-to-understand introductory resource for people interested but not experienced with compilers.</p>
<h2>What You Will End Up With</h2>
<p>If you follow this article, you should end up with a language that can define functions, call functions, define variables, assign data to variables and perform basic math operations. It will support two basic types, doubles and integers. Some of the functionality is unimplemented, so you can have the satisfaction of actually implementing some of this stuff yourself and get the hang of writing a compiler with a little help.</p>
<h2>Let&rsquo;s Get Some Questions Out of the Way</h2>
<h3>1. What languages do I need to know?</h3>
<p>The tools we&rsquo;ll be using are C/C++ based. LLVM is specifically C++ and our toy language will follow suit since there are some niceties of OOP and the STL (C++&rsquo;s stdlib) that make for fewer lines of code. In addition to C, both Lex and Bison have their own syntax which may seem daunting at first, but I&rsquo;ll try to explain as much as possible. The grammar we&rsquo;re dealing with is actually very tiny (~100 LOC), so it should be feasible.</p>
<h3>2. Is this really complicated?</h3>
<p>Yes and no. There is plenty of stuff going on here and it might seem scary at first, but honestly, this is about as simple as it gets. We will also use a lot of tools to abstract the layers of complexity and make it manageable.</p>
<h3>3. How long will it take?</h3>
<p>What we will be building took me about 3 days of toying with, but I have a few failed attempts under my belt and those do have impact on my comprehension. Of course, this will be a &ldquo;follow me&rdquo; kind of deal, so it should be much shorter for you. To understand completely everything might take a little longer, but you should be able to run through all of this stuff (and hopefully understand a good amount of it) in an afternoon.</p>
<p>So, if you&rsquo;re ready, let&rsquo;s get started.</p>
<h2>The Basic Compiler Recipe</h2>
<p>Although you should already pretty much know this, a compiler is really a grouping of three to four components (there are some more sub-components) where data is fed from one to the next in a pipeline fashion. We will be using a different tool to help us build out each of these components. Here is a diagram of each step and the tool we will use:</p>
<p><img alt="Compiler Pipeline" src="https://gnuu.org/2009/09/18/writing-your-own-toy-compiler/../../../../uploads/2009/09/pipeline.png?resize=620%2C76" data-recalc-dims="1"></p>
<p>You&rsquo;ll notice the linker step is greyed out. Our toy language won&rsquo;t be supporting compile time linking (plus most languages don&rsquo;t do compile time linking anymore anyway). To do our lexing we will be using the open source tool Lex, mostly available these days as <a href="http://flex.sourceforge.net/">Flex</a>. Lexing usually goes hand in hand with semantic parsing, which we&rsquo;ll be performing with the help of Yacc, better known as <a href="https://www.gnu.org/software/bison/">Bison</a>. Finally, once semantic parsing is done we can walk over our AST and generate our &ldquo;bytecode&rdquo;, or machine code. For this, we&rsquo;ll be using <a href="https://llvm.org">LLVM</a>, which technically generates intermediate bytecode, but we will be using LLVM&rsquo;s JIT (Just In Time) compilation to execute this bytecode on our machine.</p>
<p>To summarize, the steps are as follows:</p>
<ol>
<li>
<strong>Lexical Analysis with <em>Flex</em></strong>: Split input data into a set of tokens (identifiers, keywords, numbers, brackets, braces, etc.)</li>
<li>
<strong>Semantic Parsing with <em>Bison</em></strong>: Generate an AST while parsing the tokens. Bison will do most of the legwork here, we just need to define our AST.</li>
<li>
<strong>Assembly with <em>LLVM</em></strong>: This is where we walk over our AST and generate byte/machine code for each node. As crazy as it sounds, this is probably the <em>easiest</em> step.</li>
</ol>
<p>Before moving too far along, you should probably consider installing Flex, Bison and LLVM, if you haven&rsquo;t already. We&rsquo;re going to need them pretty soon.</p>
<h2>Defining Our Grammar</h2>
<p>Our grammar is naturally the most central part of our language. From our grammar we inherently allow or disallow functionality. For our toy compiler, we will be using a standard C-like syntax because it&rsquo;s familiar and simple to parse. The canonical example of our toy language will be the following code:</p>
<pre><code><span>int</span> <span>do_math</span>(<span>int</span> <span>a</span>) {
  <span>int</span> <span>x</span> = <span>a</span> * <span>5</span> + <span>3</span>
}

<span>do_math</span>(<span>10</span>)
</code></pre>
<p>Looks simple enough. We can see it&rsquo;s typed the same way C is, but there are no semicolons separating statements. You&rsquo;ll also notice there is no &ldquo;return&rdquo; statement in our grammar; this is something you can implement on your own.</p>
<p>There&rsquo;s also no mechanism to print any results or anything. We will verify the correctness of our programs by inspecting the <em>very pretty</em> instruction listing that LLVM can print of the bytecode we&rsquo;ve compiled, but more on that later.</p>
<h2>Step 1. Lexical Analysis with Flex</h2>
<p>This is the simplest step. Given our grammar, we need to break down our input into a list of known tokens. As mentioned before, our grammar has very basic tokens: identifiers, numbers (integers and floats), the mathematical operators, parentheses and braces. Our lex file &ldquo;tokens.l&rdquo;, which has a somewhat specialized grammar, is simply defined as:</p>
<p>Listing of tokens.l:</p>
<pre><code>%{
<span>#include &lt;string&gt;</span>
<span>#include "node.h"</span>
<span>#include "parser.hpp"</span>
<span>#define SAVE_TOKEN yylval.string = new std::string(yytext, yyleng)</span>
<span>#define TOKEN(t) (yylval.token = t)</span>
<span>extern</span> <span>"C"</span> <span>int</span> yywrap() { }
%}

%%

[ \t\n]                 ;
[a-zA-Z_][a-zA-Z0-<span>9</span>_]*  SAVE_TOKEN; <span>return</span> TIDENTIFIER;
[<span>0</span>-<span>9</span>]+.[<span>0</span>-<span>9</span>]*           SAVE_TOKEN; <span>return</span> TDOUBLE;
[<span>0</span>-<span>9</span>]+                  SAVE_TOKEN; <span>return</span> TINTEGER;
<span>"="</span>                     <span>return</span> TOKEN(TEQUAL);
<span>"=="</span>                    <span>return</span> TOKEN(TCEQ);
<span>"!="</span>                    <span>return</span> TOKEN(TCNE);
<span>"&lt;"</span>                     <span>return</span> TOKEN(TCLT);
<span>"&lt;="</span>                    <span>return</span> TOKEN(TCLE);
<span>"&gt;"</span>                     <span>return</span> TOKEN(TCGT);
<span>"&gt;="</span>                    <span>return</span> TOKEN(TCGE);
<span>"("</span>                     <span>return</span> TOKEN(TLPAREN);
<span>")"</span>                     <span>return</span> TOKEN(TRPAREN);
<span>"{"</span>                     <span>return</span> TOKEN(TLBRACE);
<span>"}"</span>                     <span>return</span> TOKEN(TRBRACE);
<span>"."</span>                     <span>return</span> TOKEN(TDOT);
<span>","</span>                     <span>return</span> TOKEN(TCOMMA);
<span>"+"</span>                     <span>return</span> TOKEN(TPLUS);
<span>"-"</span>                     <span>return</span> TOKEN(TMINUS);
<span>"*"</span>                     <span>return</span> TOKEN(TMUL);
<span>"/"</span>                     <span>return</span> TOKEN(TDIV);
.                       printf(<span>"Unknown token!n"</span>); yyterminate();

%%
</code></pre>
<p>The first section declares some specialized C code. We use a &ldquo;SAVE_TOKEN&rdquo; macro to keep the text of identifiers and numbers somewhere safe (instead of just the token itself), since Bison won&rsquo;t have access to our &lsquo;yytext&rsquo; variable. The first token tells us to skip all whitespace. You&rsquo;ll also notice that we have some equality comparison tokens and such. Those are unimplemented for now, feel free to support them in your toy compiler!</p>
<p>So all we&rsquo;re doing here is defining the tokens and their symbolic names. These symbols (TIDENTIFIER, etc.) will become &ldquo;terminal symbols&rdquo; in our grammar. We&rsquo;re just returning them, but we&rsquo;ve never defined them. <strong>Where are they defined?</strong> Why, in the bison grammar, of course. The parser.hpp file we&rsquo;ve included will be generated by bison, and all the tokens inside it will be generated and available for use.</p>
<p>We run Flex on this tokens.l file to generate our &ldquo;tokens.cpp&rdquo; file, which will be compiled alongside our parser and provide the yylex() function that recognizes all of these tokens. We will run this command later though, because we need to generate that header file from bison first!</p>
<h2>Step 2. Semantic Parsing with Bison</h2>
<p>This is the challenging part of our mission. Generating an accurate, unambiguous grammar is not simple and takes a little bit of practice. Fortunately, our grammar is both simple and, for your benefit, mostly completed. Before we get into implementing our grammar, though, we need to talk design for a bit.</p>
<h3>Designing the AST</h3>
<p>The end product of semantic parsing is an AST. As we will see, Bison is quite elegantly optimized to generate AST&rsquo;s; it&rsquo;s really only a matter of plugging our nodes into the grammar.</p>
<p>The same way a string such as &ldquo;int x&rdquo; represents our language in text form, our AST is what represents our language in memory (before it is assembled). As such, we need to build up the data structures we will be using before we have the chance of plugging them into our grammar file. This process is pretty straightforward because we are basically creating a structure for every semantic that can be expressed by our language. Method calls, method declarations, variable declarations, references, these are all candidates for AST nodes. A complete diagram of the nodes in our language is as follows:</p>
<p><img border="0" alt="Our Toy Language AST" src="https://gnuu.org/2009/09/18/writing-your-own-toy-compiler/../../../../uploads/2009/09/ClassDiagram.png?resize=640%2C505" data-recalc-dims="1"></p>
<p>The C++ that represents the above specifications is:</p>
<p>Listing of node.h:</p>
<pre><code><span>#include &lt;iostream&gt;</span>
<span>#include &lt;vector&gt;</span>
<span>#include &lt;llvm/Value.h&gt;</span>

<span>class</span> CodeGenContext;
<span>class</span> NStatement;
<span>class</span> NExpression;
<span>class</span> NVariableDeclaration;

<span>typedef</span> <span>std</span>::<span><span>vector</span>&lt;NStatement*&gt;</span> StatementList;
<span>typedef</span> <span>std</span>::<span><span>vector</span>&lt;NExpression*&gt;</span> ExpressionList;
<span>typedef</span> <span>std</span>::<span><span>vector</span>&lt;NVariableDeclaration*&gt;</span> VariableList;

<span>class</span> Node {
<span>public</span>:
    <span>virtual</span> ~Node() {}
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context) { }
};

<span>class</span> NExpression : <span>public</span> Node {
};

<span>class</span> NStatement : <span>public</span> Node {
};

<span>class</span> NInteger : <span>public</span> NExpression {
<span>public</span>:
    <span>long</span> <span>long</span> value;
    NInteger(<span>long</span> <span>long</span> value) : value(value) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NDouble : <span>public</span> NExpression {
<span>public</span>:
    <span>double</span> value;
    NDouble(<span>double</span> value) : value(value) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NIdentifier : <span>public</span> NExpression {
<span>public</span>:
    <span>std</span>::<span>string</span> name;
    NIdentifier(<span>const</span> <span>std</span>::<span>string</span>&amp; name) : name(name) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NMethodCall : <span>public</span> NExpression {
<span>public</span>:
    <span>const</span> NIdentifier&amp; id;
    ExpressionList arguments;
    NMethodCall(<span>const</span> NIdentifier&amp; id, ExpressionList&amp; arguments) :
        id(id), arguments(arguments) { }
    NMethodCall(<span>const</span> NIdentifier&amp; id) : id(id) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NBinaryOperator : <span>public</span> NExpression {
<span>public</span>:
    <span>int</span> op;
    NExpression&amp; lhs;
    NExpression&amp; rhs;
    NBinaryOperator(NExpression&amp; lhs, <span>int</span> op, NExpression&amp; rhs) :
        lhs(lhs), rhs(rhs), op(op) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NAssignment : <span>public</span> NExpression {
<span>public</span>:
    NIdentifier&amp; lhs;
    NExpression&amp; rhs;
    NAssignment(NIdentifier&amp; lhs, NExpression&amp; rhs) : 
        lhs(lhs), rhs(rhs) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NBlock : <span>public</span> NExpression {
<span>public</span>:
    StatementList statements;
    NBlock() { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NExpressionStatement : <span>public</span> NStatement {
<span>public</span>:
    NExpression&amp; expression;
    NExpressionStatement(NExpression&amp; expression) : 
        expression(expression) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NVariableDeclaration : <span>public</span> NStatement {
<span>public</span>:
    <span>const</span> NIdentifier&amp; type;
    NIdentifier&amp; id;
    NExpression *assignmentExpr;
    NVariableDeclaration(<span>const</span> NIdentifier&amp; type, NIdentifier&amp; id) :
        type(type), id(id) { }
    NVariableDeclaration(<span>const</span> NIdentifier&amp; type, NIdentifier&amp; id, NExpression *assignmentExpr) :
        type(type), id(id), assignmentExpr(assignmentExpr) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};

<span>class</span> NFunctionDeclaration : <span>public</span> NStatement {
<span>public</span>:
    <span>const</span> NIdentifier&amp; type;
    <span>const</span> NIdentifier&amp; id;
    VariableList arguments;
    NBlock&amp; block;
    NFunctionDeclaration(<span>const</span> NIdentifier&amp; type, <span>const</span> NIdentifier&amp; id, 
            <span>const</span> VariableList&amp; arguments, NBlock&amp; block) :
        type(type), id(id), arguments(arguments), block(block) { }
    <span>virtual</span> llvm::Value* codeGen(CodeGenContext&amp; context);
};
</code></pre>
<p>Again, fairly straightforward. We&rsquo;re not using any getters or setters here, just public data members; there&rsquo;s really no need for data hiding. Ignore the codeGen method for now. It will come in handy later, when we want to export our AST as LLVM bytecode.</p>
<h3>Back to Bison</h3>
<p>Moo. I mean, whatever Bison say. Here&rsquo;s a short tangent:</p>
<p>Anyway, here comes the most complex part and by the same token, the hardest to explain. It&rsquo;s not technically complicated, but I&rsquo;d have to spend a while discussing the details of Bison&rsquo;s syntax, so let&rsquo;s take a lot of it for granted right now. Realize this, though: we&rsquo;ll be declaring the makeup of each of our valid statements and expressions (the ones representing the nodes we&rsquo;ve defined above) using terminal and non-terminal symbols, basically like a BNF grammar. The syntax is also similar:</p>
<pre><code><span>if_stmt</span> <span>:</span> <span>IF</span> <span>'('</span> <span>condition</span> <span>')'</span> <span>block</span> { <span>/* do stuff when this rule is encountered */</span> }
        | <span>IF</span> <span>'('</span> <span>condition</span> <span>')'</span>       { ... }
        ;
</code></pre>
<p>The above would define an if statement (<em>if</em> we supported it). The real difference is that with each grammar comes a set of actions (inside the braces) which are executed after it is recognized. This is done recursively over the symbols in leaf-to-root order, where each non terminal is eventually merged into one big tree. The &ldquo;$$&rdquo; symbol you will be seeing represents the current root node of each leaf. Furthermore, &ldquo;$1&rdquo; represents the leaf for the 1st symbol in the rule. In the above example, the &ldquo;$$&rdquo; we assigned from our &ldquo;condition&rdquo; rule would be available as &ldquo;$3&rdquo; in our &ldquo;if_stmt&rdquo; rule. It might start becoming clear how our AST ties into this. We will basically be assigning a node to &ldquo;$$&rdquo; at each of our rules which will eventually be merged into our final AST. If not, don&rsquo;t worry. The Bison part of our toy language is again, the most complex portion of our language. It might take a while to sink in. Besides, you haven&rsquo;t even seen the code yet, so, here it is:</p>
<p>Listing of parser.y:</p>
<pre><code>%{
    <span>#include "node.h"</span>
    NBlock *programBlock; 

    <span>extern</span> <span>int</span> yylex();
    <span>void</span> yyerror(<span>const</span> <span>char</span> *s) { printf(<span>"ERROR: %sn"</span>, s); }
%}

%<span>union</span> {
    Node *node;
    NBlock *block;
    NExpression *expr;
    NStatement *stmt;
    NIdentifier *ident;
    NVariableDeclaration *var_decl;
    <span>std</span>::<span><span>vector</span>&lt;NVariableDeclaration*&gt;</span> *varvec;
    <span>std</span>::<span><span>vector</span>&lt;NExpression*&gt;</span> *exprvec;
    <span>std</span>::<span>string</span> *<span>string</span>;
    <span>int</span> token;
}

%token &lt;<span>string</span>&gt; TIDENTIFIER TINTEGER TDOUBLE
%token &lt;token&gt; TCEQ TCNE TCLT TCLE TCGT TCGE TEQUAL
%token &lt;token&gt; TLPAREN TRPAREN TLBRACE TRBRACE TCOMMA TDOT
%token &lt;token&gt; TPLUS TMINUS TMUL TDIV

%type &lt;ident&gt; ident
%type &lt;expr&gt; numeric expr 
%type &lt;varvec&gt; func_decl_args
%type &lt;exprvec&gt; call_args
%type &lt;block&gt; program stmts block
%type &lt;stmt&gt; stmt var_decl func_decl
%type &lt;token&gt; comparison

%left TPLUS TMINUS
%left TMUL TDIV

%start program

%%

program : stmts { programBlock = $<span>1</span>; }
        ;
        
stmts : stmt { $$ = <span>new</span> NBlock(); $$-&gt;statements.push_back($&lt;stmt&gt;<span>1</span>); }
      | stmts stmt { $<span>1</span>-&gt;statements.push_back($&lt;stmt&gt;<span>2</span>); }
      ;

stmt : var_decl | func_decl
     | expr { $$ = <span>new</span> NExpressionStatement(*$<span>1</span>); }
     ;

block : TLBRACE stmts TRBRACE { $$ = $<span>2</span>; }
      | TLBRACE TRBRACE { $$ = <span>new</span> NBlock(); }
      ;

var_decl : ident ident { $$ = <span>new</span> NVariableDeclaration(*$<span>1</span>, *$<span>2</span>); }
         | ident ident TEQUAL expr { $$ = <span>new</span> NVariableDeclaration(*$<span>1</span>, *$<span>2</span>, $<span>4</span>); }
         ;
        
func_decl : ident ident TLPAREN func_decl_args TRPAREN block 
            { $$ = <span>new</span> NFunctionDeclaration(*$<span>1</span>, *$<span>2</span>, *$<span>4</span>, *$<span>6</span>); <span>delete</span> $<span>4</span>; }
          ;
    
func_decl_args :   { $$ = <span>new</span> VariableList(); }
          | var_decl { $$ = <span>new</span> VariableList(); $$-&gt;push_back($&lt;var_decl&gt;<span>1</span>); }
          | func_decl_args TCOMMA var_decl { $<span>1</span>-&gt;push_back($&lt;var_decl&gt;<span>3</span>); }
          ;

ident : TIDENTIFIER { $$ = <span>new</span> NIdentifier(*$<span>1</span>); <span>delete</span> $<span>1</span>; }
      ;

numeric : TINTEGER { $$ = <span>new</span> NInteger(atol($<span>1</span>-&gt;c_str())); <span>delete</span> $<span>1</span>; }
        | TDOUBLE { $$ = <span>new</span> NDouble(atof($<span>1</span>-&gt;c_str())); <span>delete</span> $<span>1</span>; }
        ;
    
expr : ident TEQUAL expr { $$ = <span>new</span> NAssignment(*$&lt;ident&gt;<span>1</span>, *$<span>3</span>); }
     | ident TLPAREN call_args TRPAREN { $$ = <span>new</span> NMethodCall(*$<span>1</span>, *$<span>3</span>); <span>delete</span> $<span>3</span>; }
     | ident { $&lt;ident&gt;$ = $<span>1</span>; }
     | numeric
     | expr comparison expr { $$ = <span>new</span> NBinaryOperator(*$<span>1</span>, $<span>2</span>, *$<span>3</span>); }
     | TLPAREN expr TRPAREN { $$ = $<span>2</span>; }
     ;
    
call_args :   { $$ = <span>new</span> ExpressionList(); }
          | expr { $$ = <span>new</span> ExpressionList(); $$-&gt;push_back($<span>1</span>); }
          | call_args TCOMMA expr  { $<span>1</span>-&gt;push_back($<span>3</span>); }
          ;

comparison : TCEQ | TCNE | TCLT | TCLE | TCGT | TCGE 
           | TPLUS | TMINUS | TMUL | TDIV
           ;

%%
</code></pre>
<h3>Generating Our Code</h3>
<p>So we have our &ldquo;tokens.l&rdquo; file for Flex and our &ldquo;parser.y&rdquo; file for Bison. To generate our C++ source files from these definition files we need to pass them through the tools. Note that Bison will also be creating a &ldquo;parser.hpp&rdquo; header file for Flex; it does this because of the <code>&ndash;d</code> switch, which separates our token declarations from the source so we can include and use those tokens elsewhere. The following commands should create our parser.cpp, parser.hpp and tokens.cpp source files.</p>
<pre><code><span>$ </span><span>bison</span> -<span>d</span> -<span>o</span> <span>parser</span>.<span>cpp</span> <span>parser</span>.<span>y</span>
<span>$ </span><span>lex</span> -<span>o</span> <span>tokens</span>.<span>cpp</span> <span>tokens</span>.<span>l</span>
</code></pre>
<p>If everything went well we should now have 2 out of 3 parts of our compiler. If you want to test this, create a short main function in a main.cpp file:</p>
<p>Listing of main.cpp:</p>
<pre><code>

<span>extern</span> <span>NBlock</span>* <span>programBlock</span>;
<span>extern</span> <span>int</span> <span>yyparse</span>();

<span>int</span> <span>main</span>(<span>int</span> <span>argc</span>, <span>char</span> **<span>argv</span>)
{
    <span>yyparse</span>();
    <span>std</span><span>:</span><span>:<span>cout</span></span> &lt;&lt; <span>programBlock</span> &lt;&lt; <span>std</span><span>:</span><span>:<span>endl</span></span>;
    <span><span>return</span></span> <span>0</span>;
}
</code></pre>
<p>You can then compile your source files:</p>
<pre><code><span>$ </span><span>g</span>++ -<span>o</span> <span>parser</span> <span>parser</span>.<span>cpp</span> <span>tokens</span>.<span>cpp</span> <span>main</span>.<span>cpp</span>
</code></pre>
<p>You will need to have installed LLVM by now for the <code>llvm::Value</code> reference in &ldquo;node.h&rdquo;. If you don't want to go that far just yet, you can comment out the codeGen() methods in node.h to test just your lexer/parser combo.</p>
<p>If all goes well you should now have a &ldquo;parser&rdquo; binary that takes source in stdin and prints out a hopefully nonzero address representing the root node of our AST in memory.</p>
<h2>Step 3. Assembling the AST with LLVM</h2>
<p>The next step in a compiler is to naturally take this AST and turn it into machine code. This means converting each semantic node into the equivalent machine instruction. LLVM makes this very easy for us, because it abstracts the actual instructions to something that is similar to an AST. This means all we&rsquo;re really doing is translating from one AST to another.</p>
<p>You can imagine that this process will involve us walking over our AST from our root node, and for each node we emit bytecode. This is where the <code>codeGen</code> method that we defined for our nodes comes in handy. For example, when we visit an <code>NBlock</code> node (semantically representing a collection of statements), we call <code>codeGen</code> on each statement in the list. It looks like this:</p>
<pre><code>Value* NBlock::codeGen(CodeGenContext&amp; context)
{
    StatementList::const_iterator it;
    Value *last = NULL;
    <span>for</span> (it = statements.begin(); it != statements.end(); it++) {
        <span>std</span>::<span>cout</span> &lt;&lt; <span>"Generating code for "</span> &lt;&lt; <span>typeid</span>(**it).name() &lt;&lt; <span>std</span>::endl;
        last = (**it).codeGen(context);
    }
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating block"</span> &lt;&lt; <span>std</span>::endl;
    <span>return</span> last;
}
</code></pre>
<p>We implement this method for all of our nodes, then call it as we go down the tree, implicitly walking us through our entire AST. We keep a new <code>CodeGenContext</code> class around to tell us where to emit our bytecode.</p>
<h3>A Big Caveat About LLVM</h3>
<p>One of the downsides of LLVM is that it&rsquo;s <em>really</em> hard to find useful documentation. Their online tutorial and other such docs are wildly out of date, and there&rsquo;s barely any information for the C++ API unless you really dig for it. If you installed LLVM on your own, check the &lsquo;docs&rsquo; as it has &ldquo;more&rdquo; up to date documentation.</p>
<p>I&rsquo;ve found the best way to learn LLVM is by example. There are some quick examples of programmatically generating bytecode in the &lsquo;examples&rsquo; directory of the LLVM archive as well, and there&rsquo;s also <a href="https://llvm.org/demo/">the LLVM live demo site</a> which can emit C++ API code for a C program as input. This is a great way to find out what instructions something like &ldquo;int x = 5;&rdquo; will emit. I used the demo tool to implement most of the nodes.</p>
<p>Without further ado, I&rsquo;ll be listing both the codegen.h and codegen.cpp files below.</p>
<p>Listing of codegen.h:</p>
<pre><code><span>#include &lt;stack&gt;</span>
<span>#include &lt;llvm/Module.h&gt;</span>
<span>#include &lt;llvm/Function.h&gt;</span>
<span>#include &lt;llvm/Type.h&gt;</span>
<span>#include &lt;llvm/DerivedTypes.h&gt;</span>
<span>#include &lt;llvm/LLVMContext.h&gt;</span>
<span>#include &lt;llvm/PassManager.h&gt;</span>
<span>#include &lt;llvm/Instructions.h&gt;</span>
<span>#include &lt;llvm/CallingConv.h&gt;</span>
<span>#include &lt;llvm/Bitcode/ReaderWriter.h&gt;</span>
<span>#include &lt;llvm/Analysis/Verifier.h&gt;</span>
<span>#include &lt;llvm/Assembly/PrintModulePass.h&gt;</span>
<span>#include &lt;llvm/Support/IRBuilder.h&gt;</span>
<span>#include &lt;llvm/ModuleProvider.h&gt;</span>
<span>#include &lt;llvm/Target/TargetSelect.h&gt;</span>
<span>#include &lt;llvm/ExecutionEngine/GenericValue.h&gt;</span>
<span>#include &lt;llvm/ExecutionEngine/JIT.h&gt;</span>
<span>#include &lt;llvm/Support/raw_ostream.h&gt;</span>

<span>using</span> <span>namespace</span> llvm;

<span>class</span> NBlock;

<span>class</span> CodeGenBlock {
<span>public</span>:
    BasicBlock *block;
    <span>std</span>::<span><span>map</span>&lt;<span>std</span>::<span>string</span>, Value*&gt;</span> locals;
};

<span>class</span> CodeGenContext {
    <span>std</span>::<span><span>stack</span>&lt;CodeGenBlock *&gt;</span> blocks;
    Function *mainFunction;

<span>public</span>:
    Module *module;
    CodeGenContext() { module = <span>new</span> Module(<span>"main"</span>, getGlobalContext()); }
    
    <span>void</span> generateCode(NBlock&amp; root);
    GenericValue runCode();
    <span>std</span>::<span><span>map</span>&lt;<span>std</span>::<span>string</span>, Value*&gt;</span>&amp; locals() { <span>return</span> blocks.top()-&gt;locals; }
    BasicBlock *currentBlock() { <span>return</span> blocks.top()-&gt;block; }
    <span>void</span> pushBlock(BasicBlock *block) { blocks.push(<span>new</span> CodeGenBlock()); blocks.top()-&gt;block = block; }
    <span>void</span> popBlock() { CodeGenBlock *top = blocks.top(); blocks.pop(); <span>delete</span> top; }
};
</code></pre>
<p>Listing of codegen.cpp</p>
<pre><code><span>#include "node.h"</span>
<span>#include "codegen.h"</span>
<span>#include "parser.hpp"</span>

<span>using</span> <span>namespace</span> <span>std</span>;

<span>void</span> CodeGenContext::generateCode(NBlock&amp; root)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Generating code...n"</span>;
    
    
    <span><span>vector</span>&lt;<span>const</span> Type*&gt;</span> argTypes;
    FunctionType *ftype = FunctionType::get(Type::getVoidTy(getGlobalContext()), argTypes, <span>false</span>);
    mainFunction = Function::Create(ftype, GlobalValue::InternalLinkage, <span>"main"</span>, module);
    BasicBlock *bblock = BasicBlock::Create(getGlobalContext(), <span>"entry"</span>, mainFunction, <span>0</span>);
    
    
    pushBlock(bblock);
    root.codeGen(*<span>this</span>); 
    ReturnInst::Create(getGlobalContext(), bblock);
    popBlock();
    
    
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Code is generated.n"</span>;
    PassManager pm;
    pm.add(createPrintModulePass(&amp;outs()));
    pm.run(*module);
}

GenericValue CodeGenContext::runCode() {
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Running code...n"</span>;
    ExistingModuleProvider *mp = <span>new</span> ExistingModuleProvider(module);
    ExecutionEngine *ee = ExecutionEngine::create(mp, <span>false</span>);
    <span><span>vector</span>&lt;GenericValue&gt;</span> noargs;
    GenericValue v = ee-&gt;runFunction(mainFunction, noargs);
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Code was run.n"</span>;
    <span>return</span> v;
}

<span>static</span> <span>const</span> Type *typeOf(<span>const</span> NIdentifier&amp; type) 
{
    <span>if</span> (type.name.compare(<span>"int"</span>) == <span>0</span>) {
        <span>return</span> Type::getInt64Ty(getGlobalContext());
    }
    <span>else</span> <span>if</span> (type.name.compare(<span>"double"</span>) == <span>0</span>) {
        <span>return</span> Type::getDoubleTy(getGlobalContext());
    }
    <span>return</span> Type::getVoidTy(getGlobalContext());
}

Value* NInteger::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating integer: "</span> &lt;&lt; value &lt;&lt; <span>std</span>::endl;
    <span>return</span> ConstantInt::get(Type::getInt64Ty(getGlobalContext()), value, <span>true</span>);
}

Value* NDouble::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating double: "</span> &lt;&lt; value &lt;&lt; <span>std</span>::endl;
    <span>return</span> ConstantFP::get(Type::getDoubleTy(getGlobalContext()), value);
}

Value* NIdentifier::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating identifier reference: "</span> &lt;&lt; name &lt;&lt; <span>std</span>::endl;
    <span>if</span> (context.locals().find(name) == context.locals().end()) {
        <span>std</span>::<span>cerr</span> &lt;&lt; <span>"undeclared variable "</span> &lt;&lt; name &lt;&lt; <span>std</span>::endl;
        <span>return</span> NULL;
    }
    <span>return</span> <span>new</span> LoadInst(context.locals()[name], <span>""</span>, <span>false</span>, context.currentBlock());
}

Value* NMethodCall::codeGen(CodeGenContext&amp; context)
{
    Function *function = context.module-&gt;getFunction(id.name.c_str());
    <span>if</span> (function == NULL) {
        <span>std</span>::<span>cerr</span> &lt;&lt; <span>"no such function "</span> &lt;&lt; id.name &lt;&lt; <span>std</span>::endl;
    }
    <span>std</span>::<span><span>vector</span>&lt;Value*&gt;</span> args;
    ExpressionList::const_iterator it;
    <span>for</span> (it = arguments.begin(); it != arguments.end(); it++) {
        args.push_back((**it).codeGen(context));
    }
    CallInst *call = CallInst::Create(function, args.begin(), args.end(), <span>""</span>, context.currentBlock());
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating method call: "</span> &lt;&lt; id.name &lt;&lt; <span>std</span>::endl;
    <span>return</span> call;
}

Value* NBinaryOperator::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating binary operation "</span> &lt;&lt; op &lt;&lt; <span>std</span>::endl;
    Instruction::BinaryOps instr;
    <span>switch</span> (op) {
        <span>case</span> TPLUS:     instr = Instruction::Add; <span>goto</span> math;
        <span>case</span> TMINUS:    instr = Instruction::Sub; <span>goto</span> math;
        <span>case</span> TMUL:      instr = Instruction::Mul; <span>goto</span> math;
        <span>case</span> TDIV:      instr = Instruction::SDiv; <span>goto</span> math;
                
        
    }

    <span>return</span> NULL;
math:
    <span>return</span> BinaryOperator::Create(instr, lhs.codeGen(context), 
        rhs.codeGen(context), <span>""</span>, context.currentBlock());
}

Value* NAssignment::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating assignment for "</span> &lt;&lt; lhs.name &lt;&lt; <span>std</span>::endl;
    <span>if</span> (context.locals().find(lhs.name) == context.locals().end()) {
        <span>std</span>::<span>cerr</span> &lt;&lt; <span>"undeclared variable "</span> &lt;&lt; lhs.name &lt;&lt; <span>std</span>::endl;
        <span>return</span> NULL;
    }
    <span>return</span> <span>new</span> StoreInst(rhs.codeGen(context), context.locals()[lhs.name], <span>false</span>, context.currentBlock());
}

Value* NBlock::codeGen(CodeGenContext&amp; context)
{
    StatementList::const_iterator it;
    Value *last = NULL;
    <span>for</span> (it = statements.begin(); it != statements.end(); it++) {
        <span>std</span>::<span>cout</span> &lt;&lt; <span>"Generating code for "</span> &lt;&lt; <span>typeid</span>(**it).name() &lt;&lt; <span>std</span>::endl;
        last = (**it).codeGen(context);
    }
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating block"</span> &lt;&lt; <span>std</span>::endl;
    <span>return</span> last;
}

Value* NExpressionStatement::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Generating code for "</span> &lt;&lt; <span>typeid</span>(expression).name() &lt;&lt; <span>std</span>::endl;
    <span>return</span> expression.codeGen(context);
}

Value* NVariableDeclaration::codeGen(CodeGenContext&amp; context)
{
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating variable declaration "</span> &lt;&lt; type.name &lt;&lt; <span>" "</span> &lt;&lt; id.name &lt;&lt; <span>std</span>::endl;
    AllocaInst *alloc = <span>new</span> AllocaInst(typeOf(type), id.name.c_str(), context.currentBlock());
    context.locals()[id.name] = alloc;
    <span>if</span> (assignmentExpr != NULL) {
        NAssignment assn(id, *assignmentExpr);
        assn.codeGen(context);
    }
    <span>return</span> alloc;
}

Value* NFunctionDeclaration::codeGen(CodeGenContext&amp; context)
{
    <span><span>vector</span>&lt;<span>const</span> Type*&gt;</span> argTypes;
    VariableList::const_iterator it;
    <span>for</span> (it = arguments.begin(); it != arguments.end(); it++) {
        argTypes.push_back(typeOf((**it).type));
    }
    FunctionType *ftype = FunctionType::get(typeOf(type), argTypes, <span>false</span>);
    Function *function = Function::Create(ftype, GlobalValue::InternalLinkage, id.name.c_str(), context.module);
    BasicBlock *bblock = BasicBlock::Create(getGlobalContext(), <span>"entry"</span>, function, <span>0</span>);

    context.pushBlock(bblock);

    <span>for</span> (it = arguments.begin(); it != arguments.end(); it++) {
        (**it).codeGen(context);
    }
    
    block.codeGen(context);
    ReturnInst::Create(getGlobalContext(), bblock);

    context.popBlock();
    <span>std</span>::<span>cout</span> &lt;&lt; <span>"Creating function: "</span> &lt;&lt; id.name &lt;&lt; <span>std</span>::endl;
    <span>return</span> function;
}
</code></pre>
<p>There is certainly a great deal to take in here, however this is the part where you should start exploring on your own. I only have a couple of notes:</p>
<ul>
<li>We use a &ldquo;stack&rdquo; of blocks in our CodeGenContext class to keep the last entered block (because instructions are added to blocks)</li>
<li>We also use this stack to keep a <a href="https://en.wikipedia.org/wiki/Symbol_table">symbol table</a> of the local variables in each block.</li>
<li>Our toy program only knows about variables in its own scope. To support the idea of &ldquo;global&rdquo; contexts you&rsquo;d need to search upwards through each block in our stack until you found a match to the symbol (rather than simply searching the top symbol table).</li>
<li>Before entering a block we should push the block and when leaving it we should pop it.</li>
</ul>
<p>The rest of the details are all related to LLVM, and again, I&rsquo;m hardly an expert on that subject. But at this point, we have all of the code we need to compile our toy language and watch it run.</p>
<h2>Building Our Toy Language</h2>
<p>We have the code, but how to we build it? LLVM can be complicated to link, fortunately if you installed LLVM you also got an <code>llvm-config</code> binary which returns all of the compiler/linker flags you need.</p>
<p>We also need to update our main.cpp file from earlier to actually compile and run our code, this time:</p>
<p>Listing of main.cpp:</p>
<pre><code><span>#include &lt; iostream&gt;</span>
<span>#include "codegen.h"</span>
<span>#include "node.h"</span>

<span>using</span> <span>namespace</span> <span>std</span>;

<span>extern</span> <span>int</span> yyparse();
<span>extern</span> NBlock* programBlock;

<span>int</span> main(<span>int</span> argc, <span>char</span> **argv)
{
    yyparse();
    <span>std</span>::<span>cout</span> &lt;&lt; programBlock &lt;&lt; <span>std</span>::endl;

    CodeGenContext context;
    context.generateCode(*programBlock);
    context.runCode();
    
    <span>return</span> <span>0</span>;
}
</code></pre>
<p>Now all we need to do is compile:</p>
<pre><code><span>g</span>++ -<span>o</span> <span>parser</span> `<span>llvm</span>-<span>config</span> --<span>libs</span> <span>core</span> <span>jit</span> <span>native</span> --<span>cxxflags</span> --<span>ldflags</span>` *.<span>cpp</span>
</code></pre>
<p>You can also grab this Makefile to make things easier:</p>
<p>Listing of Makefile:</p>
<pre><code><span>all</span><span>:</span> <span>parser</span>

<span>clean</span><span>:</span>
    <span>rm</span> <span>parser</span>.<span>cpp</span> <span>parser</span>.<span>hpp</span> <span>parser</span> <span>tokens</span>.<span>cpp</span>

<span>parser</span>.<span>cpp</span><span>:</span> <span>parser</span>.<span>y</span>
    <span>bison</span> -<span>d</span> -<span>o</span> <span>$@</span> <span>$^</span>
    
<span>parser</span>.<span>hpp</span><span>:</span> <span>parser</span>.<span>cpp</span>

<span>tokens</span>.<span>cpp</span><span>:</span> <span>tokens</span>.<span>l</span> <span>parser</span>.<span>hpp</span>
    <span>lex</span> -<span>o</span> <span>$@</span> <span>$^</span>

<span>parser</span><span>:</span> <span>parser</span>.<span>cpp</span> <span>codegen</span>.<span>cpp</span> <span>main</span>.<span>cpp</span> <span>tokens</span>.<span>cpp</span>
    <span>g</span>++ -<span>o</span> <span>$@</span> `<span>llvm</span>-<span>config</span> --<span>libs</span> <span>core</span> <span>jit</span> <span>native</span> --<span>cxxflags</span> --<span>ldflags</span>` *.<span>cpp</span>
</code></pre>
<h2>Running Our Toy Language</h2>
<p>Hopefully everything compiled properly. At this point you should have your parser binary that spits out code. Play with it. Remember our canonical example? Let&rsquo;s see what our program does.</p>
<pre><code>$ echo <span>'int do_math(int a) { int x = a * 5 + 3 } do_math(10)'</span> | ./parser
<span>0x100a00f10</span>
Generating code...
Generating code <span>for</span> <span>20</span>NFunctionDeclaration
Creating variable declaration int a
Generating code <span>for</span> <span>20</span>NVariableDeclaration
Creating variable declaration int x
Creating assignment <span>for</span> x
Creating binary operation <span>276</span>
Creating binary operation <span>274</span>
Creating integer: <span>3</span>
Creating integer: <span>5</span>
Creating identifier reference: a
Creating block
Creating <span><span>function</span>: <span>do_math</span>
<span>Generating</span> <span>code</span> <span>for</span> 20<span>NExpressionStatement</span>
<span>Generating</span> <span>code</span> <span>for</span> 11<span>NMethodCall</span>
<span>Creating</span> <span>integer</span>: 10
<span>Creating</span> <span>method</span> <span>call</span>: <span>do_math</span>
<span>Creating</span> <span>block</span>
<span>Code</span> <span>is</span> <span>generated</span>.
; <span>ModuleID</span> = '<span>main</span>'
<span>define</span> <span>internal</span> <span>void</span> @<span>main</span><span>()</span> {</span>
entry:
    %<span>0</span> = call i64 @do_math(i64 <span>10</span>)      ;  [#uses=<span>0</span>]
    ret <span>void</span>
}

define internal i64 @do_math(i64) {
entry:
    %a = alloca i64     ;  [#uses=<span>1</span>]
    %x = alloca i64     ;  [#uses=<span>1</span>]
    %<span>1</span> = add i64 <span>5</span>, <span>3</span>   ;  [#uses=<span>1</span>]
    %<span>2</span> = load i64* %a   ;  [#uses=<span>1</span>]
    %<span>3</span> = mul i64 %<span>2</span>, %<span>1</span> ;  [#uses=<span>1</span>]
    store i64 %<span>3</span>, i64* %x
    ret <span>void</span>
}
Running code...
Code was run.
</code></pre>
<p>Isn&rsquo;t that pretty cool? I&rsquo;ll admit it&rsquo;s probably less satisfying to get something running when you just copy paste from a how to guide, but getting this running should still be interesting for you. Plus, and this is the best part, this isn&rsquo;t the end of the guide, it&rsquo;s just the beginning, because this is the part where <em>you</em> start tacking on crazy features to this lexer, parser and assembler to create a language that you can call your own. You have all the building blocks, and you should have a basic idea of how to continue on.</p>
<p>The code, by the way, is available on Github <a href="https://github.com/lsegal/my_toy_compiler">here</a>. I avoided saying this because the code was not the point of the article, but rather going over the code in detail was. If you&rsquo;re just going to grab the code it sort of defeats the purpose of writing your own compiler.</p>
<p>I realize this is a huge article, and there are probably mistakes, so if you run into problems, feel free to send me an email and I&rsquo;ll make adjustments. If you want to share some info, feel free to do that as well.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via gnuu.org https://gnuu.org<br>
<br>
November 8, 2019 at 03:06PM
